blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,687 @@
+	}
+
+	// unpack event data
+	event, err := unpackEventParam(abiEvent, log)
+	if err != nil {
+		return err
+	}
+
+	// handle different kinds of event
+	switch abiEvent.Name {
+	case "BucketTypeActivated":
+		return dirty.handleBucketTypeActivatedEvent(event, blk.Height())
+	case "BucketTypeDeactivated":
+		return dirty.handleBucketTypeDeactivatedEvent(event, blk.Height())
+	case "Staked":
+		return dirty.handleStakedEvent(event, blk.Height())
+	case "Locked":
+		return dirty.handleLockedEvent(event)
+	case "Unlocked":
+		return dirty.handleUnlockedEvent(event, blk.Height())
+	case "Unstaked":
+		return dirty.handleUnstakedEvent(event, blk.Height())
+	case "Merged":
+		return dirty.handleMergedEvent(event)
+	case "DurationExtended":
+		return dirty.handleDurationExtendedEvent(event)
+	case "AmountIncreased":
+		return dirty.handleAmountIncreasedEvent(event)
+	case "DelegateChanged":
+		return dirty.handleDelegateChangedEvent(event)
+	case "Withdrawal":
+		return dirty.handleWithdrawalEvent(event)
+	case "Transfer":
+		return dirty.handleTransferEvent(event)
+	default:
+		return nil
+	}
+}
+
+func (s *liquidStakingIndexer) loadCache() error {
+	delta := newLiquidStakingDelta()
+	// load height
+	var height uint64
+	h, err := s.kvstore.Get(_liquidStakingNS, _liquidStakingHeightKey)
+	if err != nil {
+		if !errors.Is(err, db.ErrNotExist) {
+			return err
+		}
+		height = 0
+	} else {
+		height = byteutil.BytesToUint64BigEndian(h)
+
+	}
+	delta.putHeight(height)
+
+	// load total bucket count
+	var totalBucketCount uint64
+	tbc, err := s.kvstore.Get(_liquidStakingNS, _liquidStakingTotalBucketCountKey)
+	if err != nil {
+		if !errors.Is(err, db.ErrNotExist) {
+			return err
+		}
+	} else {
+		totalBucketCount = byteutil.BytesToUint64BigEndian(tbc)
+	}
+	delta.putTotalBucketCount(totalBucketCount)
+
+	// load bucket info
+	ks, vs, err := s.kvstore.Filter(_liquidStakingBucketInfoNS, func(k, v []byte) bool { return true }, nil, nil)
+	if err != nil && !errors.Is(err, db.ErrBucketNotExist) {
+		return err
+	}
+	for i := range vs {
+		var b ContractStakingBucketInfo
+		if err := b.deserialize(vs[i]); err != nil {
+			return err
+		}
+		delta.putBucketInfo(byteutil.BytesToUint64BigEndian(ks[i]), &b)
+	}
+
+	// load bucket type
+	ks, vs, err = s.kvstore.Filter(_liquidStakingBucketTypeNS, func(k, v []byte) bool { return true }, nil, nil)
+	if err != nil && !errors.Is(err, db.ErrBucketNotExist) {
+		return err
+	}
+	for i := range vs {
+		var b ContractStakingBucketType
+		if err := b.deserialize(vs[i]); err != nil {
+			return err
+		}
+		delta.putBucketType(byteutil.BytesToUint64BigEndian(ks[i]), &b)
+	}
+	return s.cache.merge(delta)
+}
+
+func (s *liquidStakingIndexer) commit(dirty *liquidStakingDirty) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	batch, delta := dirty.finalize()
+	if err := s.kvstore.WriteBatch(batch); err != nil {
+		return err
+	}
+	if err := s.cache.merge(delta); err != nil {
+		return err
+	}
```

Comment:
should acid be satified here?

Reply:
sure, i think two changes need to do:
1. move `cache.merge` before `kvstore.WriteBatch`
2. `reloadCache` when error occured